### ISSUE TITLE
Store paths internally using `NameRef`s (or something `NameRef`-like)

### ISSUE THREAD

**Issue by jvilk-stripe** [sorbet/sorbet#4509]
### Problem

Sorbet stores file paths internally as full strings. These strings contain redundant substrings (directory names) and are thus not stored efficiently internally. Sorbet stores at least two full copies of each file path per global state: Once in a map that supports file lookup by path, and second in File objects themselves.

Although the memory overhead of these strings is minimal, constructing and manipulating them takes real runtime on large projects. Sorbet has to allocate a string for each file at startup and hash the full string into a hashmap, and it does so serially as GlobalState does not support parallel mutations.

The up-and-coming packager pass has to be directory aware, and is building support to identify what package a file belongs to. This pass has a lot of custom-made directory manipulations which take place on strings.

### Proposed solution

We should store a path internally as a `vector` of `NameRef`s. Explode the path on `/` and `NameRef` each component.

Pros:
* Path objects themselves are very small, which makes them cheap to copy (each path component is 4 bytes).
* When Sorbet is given a directory to traverse (common case), it uses a recursive strategy and is aware of the path up until that point. That means that it only needs to hash each file/folder once to convert them into NameRefs.
* Many of these names may already be used elsewhere in the project, saving a small amount of memory.
* Sorbet's file map in GlobalState can be turned into a trie, which 1) can support efficiently looking up all files in a directory (good for packager) and 2) may be more compact than a map.
* If we can't convert a path string component into a `NameRef`, then we know early that the given file doesn't/cant exist in Sorbet yet.

Cons:
* Going from a path string to a path object is O(path_components) as we 1) explode the string into components and 2) look up or insert `NameRef`s for each component. 
* With a trie, looking up a file for a path object is O(path_components) and may involve traversing multiple cache lines.
  * We don't _have_ to use a trie; we could still use a hashmap and teach it to hash path objects.
* If we use `NameRef`s, path objects are now tied to a specific `GlobalState`. But arguably `FileRef` objects are already in that state. This will mainly complicate LSP slightly, which juggles multiple `GlobalState` objects.

**Comment by jvilk-stripe** [comment]
cc @aadi-stripe

**Comment by jvilk-stripe** [comment]
@elliottt brings up the valid concern that `NameRef` opens us up to some footguns. There are three different types of `NameRef`s:

* UTF8: Simple strings. What we want.
* Constant: Constant names. Are validated internally that they are legal.
* Unique: Too hard to get into here; used to differentiate between two items named the same in an illegal fashion.

A `NameRef` can reference any one of these. We only want UTF8 names, so we may need a `PathRef` object that wraps/_only_ references a UTF8 NameRef.

**Comment by aadi-stripe** [comment]
@jvilk-stripe I think `PathRef` makes sense. Also just checking this -- are you implying that uniqueness should be by fully qualified path, meaning that...

`foo/bar/baz` should point to a unique `PathRef`, `foo/bar` should point to a unique `PathRef`, `foo/` should point to a unique `PathRef`, and so on?

**Comment by jvilk-stripe** [comment]
@aadi-stripe Nope, I'm saying that `foo/bar/baz` is exploded into three `PathRef`s (maybe `PathComponentRef`s?):

* `foo`
* `bar`
* `baz`

Each of these `PathComponentRef`s is internally a `UTF8` `NameRef`.

A `Path` is then a vector of these `NameRef`s. 

It then becomes straightforward to construct a trie using individual `PathComponentRef` (which are just 32-bit ints!) as the node values (so the path thru the trie is the full path).
